### PR TITLE
Implement Correct Usage Flags for `forest` and `forest-cli`

### PR DESCRIPTION
**Pull request by jdjaustin**
**Summary of changes**
Changes introduced in this pull request:
- Change flags available for `forest` and `forest-cli` to only those that make sense and are usable.

**Reference issue to close (if applicable)**

Closes associated tasks in issue #2076 

**Other information and links**

Task to change the config dump behavior will be covered under a separate PR.

### CODE REVIEW

File: forest/shared/src/cli/mod.rs [hunk 1]
```diff
@@ -351,31 +443,31 @@ mod tests {
...
+foo = "foo"
+[myth]
+author = 'H. P. Lovecraft'
+entities = [
+{ name = 'Cthulhu' },
+{ name = 'Azathoth' },
+{ baz = 'Dagon' },
+]
+bar = "bar"
+"#,
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:456]: This change looks unintentional.

File: forest/shared/src/cli/mod.rs [hunk 2]
```diff
@@ -204,7 +280,23 @@ impl ConfigPath {
...
+        return Some(ConfigPath::Env(PathBuf::from(s)));
+    }
+    if let Some(dir) = ProjectDirs::from("com", "ChainSafe", "Forest") {
+        let path = dir.config_dir().join("config.toml");
+        if path.exists() {
+            return Some(ConfigPath::Project(path));
+        }
+    }
+    None
+}
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:297]: `find_cli_config_path` and `find_daemon_config_path` should be merged. Or refactored such that they share their logic.

File: forest/shared/src/cli/mod.rs [hunk 3]
```diff
@@ -30,6 +30,82 @@ pub static FOREST_VERSION_STRING: Lazy<String> =
 /// CLI options
 #[derive(StructOpt, Debug)]
 pub struct CliOpts {
+    /// A TOML file containing relevant configurations
+    #[structopt(short, long)]
+    pub config: Option<String>,
+    /// Allow RPC to be active or not (default: true)
+    #[structopt(short, long)]
+    pub rpc: Option<bool>,
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:38]: Does `forest-cli` use the `rpc` flag?

File: forest/shared/src/cli/mod.rs [hunk 4]
```diff
@@ -83,10 +159,10 @@ pub struct CliOpts {
...
     /// Choose network chain to sync to
     #[structopt(
-        long,
-        default_value = "mainnet",
-        possible_values = &["mainnet", "calibnet"],
-    )]
+long,
+default_value = "mainnet",
+possible_values = &["mainnet", "calibnet"],
+)]
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:165]: This change looks unintentional.

File: forest/shared/src/cli/mod.rs [hunk 5]
```diff
@@ -104,14 +180,14 @@ pub struct CliOpts {
     // However, we cannot upgrade to version 0.9 because pretty_env_logger depends on version 0.7
     // and hasn't been updated in quite a while. See https://github.com/seanmonstar/pretty-env-logger/issues/52
     // #[structopt(
-    //     help = "Specify a filename into which logging should be appended"
+    // help = "Specify a filename into which logging should be appended"
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:183]: This change looks unintentional.

File: forest/shared/src/cli/mod.rs (hunk 3 above)
Comment (jdjaustin) [reply forest/shared/src/cli/mod.rs:38]: There is a test that executes `forest-cli --rpc true --token Azazello config dump`. I (perhaps incorrectly) assumed that the test was there for a reason.

Comment (lemmih) [reply forest/shared/src/cli/mod.rs:38]: You need to understand what that test does and then update it if necessary.

File: forest/shared/src/cli/mod.rs [hunk 6]
```diff
@@ -204,20 +280,44 @@ impl ConfigPath {
...
+            if let Ok(s) = std::env::var("FOREST_CONFIG_PATH") {
+                return Some(ConfigPath::Env(PathBuf::from(s)));
+            }
+            if let Some(dir) = ProjectDirs::from("com", "ChainSafe", "Forest") {
+                let path = dir.config_dir().join("config.toml");
+                if path.exists() {
+                    return Some(ConfigPath::Project(path));
+                }
+            }
+            None
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:318]: You've still duplicated the logic. The logic is exactly the same in both instances.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/shared/src/cli/mod.rs [hunk 7]
```diff
@@ -86,7 +162,7 @@ pub struct CliOpts {
         long,
         default_value = "mainnet",
         possible_values = &["mainnet", "calibnet"],
-    )]
+     )]
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:165]: This looks unintentional.

File: forest/shared/src/cli/mod.rs [hunk 8]
```diff
@@ -204,8 +280,8 @@ impl ConfigPath {
     }
 }
 
-fn find_config_path(opts: &CliOpts) -> Option<ConfigPath> {
-    if let Some(s) = &opts.config {
+fn find_config_path(opts: &Option<String>) -> Option<ConfigPath> {
```
Comment (lemmih) [inline forest/shared/src/cli/mod.rs:283]: Good. But `opts` is no longer a proper name for this variable.

**Review by lemmih — CHANGES_REQUESTED** [review]
Getting closer. You still need to review all the cli options, though. It's not too difficult to see if a command-line option is used or not. You can either use `git grep` or simply delete the field and see what breaks.

**Review by lemmih — APPROVED** [review]
There's a lot more that can be done to clean up the command-line handling for `forest` and `forest-cli` but this is a good start.

File: forest/shared/src/cli/mod.rs [hunk 9]
```diff
@@ -30,6 +30,66 @@ pub static FOREST_VERSION_STRING: Lazy<String> =
...
+    pub config: Option<String>,
+    /// Client JWT token to use for JSON-RPC authentication
+    #[structopt(short, long)]
+    pub token: Option<String>,
+    /// Address used for RPC. By defaults binds on localhost on port 1234.
+    #[structopt(long)]
+    pub rpc_address: Option<SocketAddr>,
+    /// Address used for metrics collection server. By defaults binds on localhost on port 6116.
+    #[structopt(long)]
+    pub metrics_address: Option<SocketAddr>,
```
Comment (LesnyRumcajs) [inline forest/shared/src/cli/mod.rs:44]: I don't think we use it anywhere in the `forest-cli` unless I'm missing something.
```suggestion
```

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]

File: forest/shared/src/cli/mod.rs [hunk 10]
```diff
@@ -30,6 +30,66 @@ pub static FOREST_VERSION_STRING: Lazy<String> =
...
+    // and hasn't been updated in quite a while. See https://github.com/seanmonstar/pretty-env-logger/issues/52
+    // #[structopt(
+    // help = "Specify a filename into which logging should be appended"
+    // )]
+    // pub log_file: Option<PathBuf>,
+}
+
+impl CliOpts {
+    pub fn to_config(&self) -> Result<(Config, Option<ConfigPath>), anyhow::Error> {
+        let path = find_config_path(&self.config);
```
[the remark this answers is not in the document]
Comment (lemmih) [reply forest/shared/src/cli/mod.rs:66]: There's no such thing as a `forest-cli` config, though, and there probably never will be.

Given that there is apparently no consensus about this, I think we should close this PR and move the discussion to an issue.

File: forest/shared/src/cli/mod.rs [hunk 11]
```diff
@@ -30,6 +30,66 @@ pub static FOREST_VERSION_STRING: Lazy<String> =
...
+    pub rpc_address: Option<SocketAddr>,
+    /// Address used for metrics collection server. By defaults binds on localhost on port 6116.
+    #[structopt(long)]
+    pub metrics_address: Option<SocketAddr>,
+    /// Choose network chain to sync to
+    #[structopt(
+long,
+default_value = "mainnet",
+possible_values = &["mainnet", "calibnet"],
+)]
```
[the remark this answers is not in the document]
Comment (jdjaustin) [reply forest/shared/src/cli/mod.rs:50]: `fmt` doesn't force it, but it does seem to be creating these random messy formatting issues. I can manually fix it though (without failing the checks).

**Comment by LesnyRumcajs** [comment]
One other thing, `DaemonOpts` should not have the `token` field. As mentioned in the comment above it, it's for the client, i.e., `forest-cli`.

File: forest/shared/src/cli/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (lemmih) [reply forest/shared/src/cli/mod.rs:66]: Command-line options are different from a configuration file. In `forest` they are the same (which is a mistake, I believe, as it leads to some unintuitive behaviour). But for `forest-cli` there is no reason to have a configuration file.

Let's discuss in a separate issue.

[the remark this answers is not in the document]
Comment (LesnyRumcajs) [reply forest/shared/src/cli/mod.rs:66]: I'm okay with having no configuration file for `forest-cli` and removing this logic. What I am not okay with is having the same logic for the configuration file between `forest` and `forest-cli`.

[the remark this answers is not in the document]
Comment (lemmih) [reply forest/shared/src/cli/mod.rs:66]: Why not? We want the `forest-cli` tool to read the configuration file for the `forest` node, don't we?

[the remark this answers is not in the document]
Comment (LesnyRumcajs) [reply forest/shared/src/cli/mod.rs:66]: I mean the same logic for loading two different files, including their default location. :)

**Pull request closed without merging**